Bricks/sync/owned_borrowed.h

```diff
@@ -0,0 +1,492 @@
+/*******************************************************************************
+The MIT License (MIT)
+
+Copyright (c) 2016 Dmitry "Dima" Korolev, <dmitry.korolev@gmail.com>.
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*******************************************************************************/
+
+// The synchronization primitive to leverage RAII and enable creating scope-based long-lived objects.
+//
+// Philosophy:
+// * The scope in which the object has been created is the master owner of this object; others are borrowers.
+// * Upon leaving the master scope, the synchronization primitive waits until all its borrowers terminate.
+// * Borrowers are under no obligation to terminate sooner or later, but they get signaled to wrap up.
+//
+// Implementation:
+//
+// * auto master = current::MakeOwned<T>( [ ... T's constructor arguments ]);
+//   Creates `master`, which will live only within its creation scope.
+//   The basic usage of `master` is `master->MemberFunctionOfT(...)`, much like with C++'s `shared_ptr<T>`.
+//
+// * Borrowed<T> borrower(master);  // Or `Borrowed<T> another_borrower(borrower);`.
```

Comment:
As for me, the inverted second part (`// Or Borrowed<T> borrower(another_borrower);`) looks a little more clear and symmetric.